### ISSUE TITLE
BUG Significant performance drop with multi-sampled FBO

### ISSUE THREAD

**Issue by hemostx** [ezQuake/ezquake-source#600]
**ezQuake version:**
Since `vid_framebuffer_multisample` was introduced.

**OS/device including version:**

W10 20H2, i5@4.2ghz, GTX 1070 472.12 (and a couple releases before that, with some DDU cleanups inbetween), STD or GLSL builds, 32 or 64 bit. CPU/GPU power settings are irrelevant.

**Describe the bug**
When enabling any level of multisampling with `vid_framebuffer_multisample`, the framerate takes a nosedive unseen with equivalent  `FBO 0 + gl_multisample` settings.

**To Reproduce**

1. Set `vid_framebuffer 1 or 2` and `vid_framebuffer_multisample 0`
2. Run a timedemo :)
3. Switch up `vid_framebuffer_multisample`
4. Run a timedemo :(

Alternatively,

1. Set `vid_framebuffer 1 or 2` and `vid_framebuffer_multisample 0` and `cl_maxfps 1000`
2. Go to a 'complex' room
3. Switch up `vid_framebuffer_multisample`
4. Watch the framerate plummet

**Expected behavior**
A similar impact as with gl_multisamples when not using custom FBOs.

**More info**
For example, simply setting it to 2 is much, much worse than simply rendering to a larger FBO and scaling with ie. `vid_framebuffer_scale 0.5`, and still heaps worse than setting `FBO 0 + gl_multisample 16`.

Tests conduced with a default-ish config (1290x1080 + polyblend off, dynamic 2 to ensure > 1000 fps in GLSL) with `timedemo FPS3` :

| MSAA Level | vid_framebuffer 0, gl_multisample | vid_framebuffer 1, vid_framebuffer_multisample |
| ---------- | --------------------------------- | ---------------------------------------------- |
| 0          | 1955                              | 1960                                           |
| 2          | 1904                              | 646                                            |
| 4          | 1895                              | 650                                            |
| 8          | 1740                              | 643                                            |
| 16         | 1740                              | 650                                            |

For reference: `FBO on + vid_framebuffer_multisample 0 + vid_framebuffer_scale 0.5` = 1803 fps.

More data measured with FrameView in benchmark mode at dm2 bigroom (same position), first without FBO, then with. Maxfps 1000 for consistency.

|                | Avg FPS  | Min FPS | Max FPS  | 90th %  | 95th %  | 99th %  | RenderPresentLatency (ms) | GPU0Clk(MHz) | GPU0 Util% | GPU NV Power (Watts) (API) | CPU Util % |
| -------------- | -------- | ------- | -------- | ------- | ------- | ------- | ------------------------- | ------------ | ---------- | -------------------------- | ---------- |
| FBO 0, MSAA 0  | 1000.057 | 803.019 | 1486.768 | 948.227 | 930.233 | 890.234 | 0.287                     | 1582         | 31         | 60                         | 25.765     |
| FBO 0, MSAA 2  | 1000.053 | 798.977 | 1420.858 | 952.29  | 934.667 | 906.125 | 0.327                     | 1582         | 35.232     | 63                         | 29.957     |
| FBO 0, MSAA 4  | 1000.05  | 735.565 | 1388.31  | 948.317 | 929.973 | 897.666 | 0.323                     | 1733         | 35.234     | 73                         | 29.551     |
| FBO 0, MSAA 8  | 1000.067 | 436.872 | 2004.008 | 946.163 | 929.368 | 888.178 | 0.426                     | 1898         | 46.406     | 96                         | 27.921     |
| FBO 0, MSAA 16 | 1000.056 | 809.389 | 1752.234 | 936.68  | 898.392 | 854.774 | 0.914                     | 1880.905     | 88.539     | 166                        | 30.177     |
| FBO 1, MSAA 0  | 1000.042 | 813.206 | 1617.861 | 942.063 | 927.988 | 883.783 | 0.28                      | 1582         | 25.21      | 58                         | 25.478     |
| FBO 1, MSAA 2  | 676.127  | 116.553 | 774.833  | 641.108 | 621.852 | 570.353 | 0.361                     | 1898         | 49.909     | 72                         | 41.234     |
| FBO 1, MSAA 4  | 674.218  | 136.199 | 800.32   | 640.451 | 622.587 | 582.479 | 0.382                     | 1898         | 50.818     | 75                         | 36.133     |
| FBO 1, MSAA 8  | 684.155  | 127.4   | 759.301  | 641.807 | 624.298 | 582.852 | 0.473                     | 1898         | 57.411     | 76                         | 37.746     |
| FBO 1, MSAA 16 | 689.838  | 130.577 | 765.111  | 650.745 | 630.12  | 575.805 | 0.778                     | 1898         | 78.371     | 113                        | 38.286     |

- RenderPresentLatency increases with MSAA level regardless of FBO use, ergo "it's not something killing frametimes".
- GPU clocks, power, usage ramps up quicker with custom FBO + MSAA.
- CPU jumps up too, but not to the point that it chokes hard enough to be the issue.
- Just enabling MSAA tanks framerates, it doesn't matter what level.
- MSAA is, visually, clearly working as expected for each level.

I went to run some CPU profiling in that same spot with vid_framebuffer 1 and toggling MSAA.
- No MSAA:
![debug_fbo1msaa0](https://user-images.githubusercontent.com/64284375/140645949-2999772e-9e72-4211-92c0-e7a7a4e188dd.PNG)
There's a lot more going on in `GL_FramebufferCreate` and `GL_FramebufferDelete` with MSAA on (I got similar figures with debug/release builds) - alas, it's that pesky *external code*.
- MSAA on:
![debug_fbo1msaa4](https://user-images.githubusercontent.com/64284375/140645979-dfb97a38-f070-4ac3-95b8-c1eb05b8b8b1.PNG)

On a different day I thought I'd seen `GL_MultiSamplingResolve` come up in the hot path,  but I couldn't get it to at the time of writing this report.

Profiling with good ol' gl_multisamples and no custom FBO shows the parent GLM_ScreenDrawStart at an innocuous 0.01%.

Poking with `vid_framebuffer_blit` did not have a measurable impact. I was under the impression that it would blit from the FBO to the main buffer anyway if the FBO is multisampled.

**Comment by Calinou** [comment]
I wonder if this can be reproduced in ioquake3's OpenGL 2 renderer. It offers both traditional and FBO MSAA, with FBO MSAA being more expensive than traditional MSAA.

**Comment by hemostx** [comment]
I was wondering the same thing @Calinou ---
## ioq3 opengl2 (they claim opengl vs opengl2)
```
FBO 0, MSAA 0: 942 fps
FBO 0, MSAA 4: 884 fps
FBO 1, MSAA 0: 932 fps
FBO 1, MSAA 4: 820 fps
```
So, a little hit, but nothing too nasty.

## quake3e OpenGL
```
Couldn't get MSAA to work with FBO 0
FBO 1, MSAA 0: 1700 fps
FBO 1, MSAA 4: 1580 fps
```
## cnq3, based on ioq3, appears to always use FBO regardless of render path
The only case where I can get a nasty drop is in the GL3 path:
```
MSAA 0 : 1140 fps
MSAA 4 : 685 fps
```
In both their GL2 and D3D11 paths, considering FBOs are always on, MSAA is a 80 fps drop.

**Comment by hemostx** [comment]
Back to evaluating this in ezquake, I stumbled upon an interesting manifestation of the issue while doing pre-stream checks
- With FBO MSAA off:
![no_fbo_msaa](https://user-images.githubusercontent.com/64284375/141657366-06522625-04dd-4777-ba4f-4f2571c771a3.PNG)
*(obs appears here, but it has no impact on that measurement)*

- With FBO MSAA enabled, framebuffer scale 1:
![fbo_msaa_on_scale1](https://user-images.githubusercontent.com/64284375/141657514-762a4cef-2449-46eb-acd9-2ead2f4d8632.PNG)

- With FBO MSAA enabled and extra framebuffer scaling because why not:
![fbo_msaa_on](https://user-images.githubusercontent.com/64284375/141657370-969776c7-d473-4fef-b7c3-0768455ac702.PNG)
The System process takes 50% of the GPU doing copy operations when the perf drop happens.

Interestingly vid_framebuffer 2 decreases that GPU copy usage, but the framerate gets ~100fps worse.

**Comment by meag** [comment]
Hi... +1 (hundred) for renderdoc, thanks for that

The framebuffer is being re-created each time because the check for "was framebuffer created with current settings" is expecting the non-ms framebuffer to have multi-sampling enabled... looking into fix now.

**Comment by hemostx** [comment]
> fixes #600 (hopefully)
Quick timedemo fps3 runs with the playing config:
|       | MSAA 0 | MSAA 4 |
| ----- | ------ | ------ |
| FBO 0 | 1540   | 1430   |
| FBO 1 | 1500   | 1400   |

🙏
